netdata/netdata
# Remove SIGSEGV and SIGABRT (ebpf.plugin)

##### Summary
Fixes #13390

This PR is rewritten parts of eBPF code to remove `SIGSEGV` and `SIGABRT` that appeared some times. 
We are also addressing another issue that is the time to close `eBPF.plugin`, when this PR is merged, it is expected that it will close quicker.
It also adds some protection against possible race conditions.
This is not solving all issues related to memory. Both `apps` and `ebpf` are not cleaning neither their link lists nor `apps groups`, this needs to be address in another PR after the issues present in this PR are fixed. To verify these issues it is necessary to run `valgrind 3.19` or newer.

##### Test Plan

1. Run `ulimit -S -c unlimited` or if you use `systemd`, install `systemd-coredump` (or something like this, I do not use `systemd`)
2. Compile this branch
3. Run Netdata and wait few minutes to load data.
4. Stop netdata and run: `ps aux | grep netdata`, `ebpf.plugin` should stop in maximum 3 seconds depending of the environment.
4. Verify presence of `core` or run `coredumpctl --list`, you should not have neither core dumps nor `sigabrt`.

##### Additional Information
It is not necessary to test on different kernels.

<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ebpf.plugin
- Can they see the change or is it an under the hood? If they can see it, where? When they stop plugins it will exit quicker and without error.
- How is the user impacted by the change? A more stable plugin.
- What are there any benefits of the change? A more stable plugin.
</details>
